# 异常过滤器

内置的**异常层**负责处理整个应用程序中的所有抛出的异常。

当捕获到未处理的异常时，最终用户将收到友好的响应。

```typescript
@Get()
async findAll() {
  throw new HttpException('Forbidden', HttpStatus.FORBIDDEN);
}
```

```json
{
    "statusCode": 403,
    "message": "Forbidden"
}
```

`HttpException` 构造函数有两个必要的参数来决定响应:

-   `response` 参数定义 `JSON` 响应体。它可以是 `string` 或 `object`，如下所述。
-   `status`参数定义`HTTP`状态代码

默认情况下，`JSON` 响应主体包含两个属性：

-   `statusCode`：默认为 `status` 参数中提供的 `HTTP` 状态代码
-   `message`:基于状态的 `HTTP` 错误的简短描述

仅覆盖 `JSON` 响应主体的消息部分，请在 `response`参数中提供一个 `string`。

要覆盖整个 `JSON` 响应主体，请在`response` 参数中传递一个`object`。 `Nest`将序列化对象，并将其作为`JSON` 响应返回。

第二个构造函数参数-`status`-是有效的 `HTTP` 状态代码。 最佳实践是使用从`@nestjs/common`导入的 `HttpStatus`枚举。

覆盖整个响应正文的实例

```typescript
@Get()
async findAll() {
  throw new HttpException({
    status: HttpStatus.FORBIDDEN,
    error: 'This is a custom message',
  }, HttpStatus.FORBIDDEN);
}
```

## 内置HTTP异常

-   `BadRequestException`
-   `UnauthorizedException`
-   `NotFoundException`
-   `ForbiddenException`
-   `NotAcceptableException`
-   `RequestTimeoutException`
-   `ConflictException`
-   `GoneException`
-   `PayloadTooLargeException`
-   `UnsupportedMediaTypeException`
-   `UnprocessableException`
-   `InternalServerErrorException`
-   `NotImplementedException`
-   `BadGatewayException`
-   `ServiceUnavailableException`
-   `GatewayTimeoutException`

## 异常过滤器

```typescript
import { ExceptionFilter, Catch, ArgumentsHost, HttpException } from '@nestjs/common';
import { Request, Response } from 'express';

@Catch(HttpException)
export class HttpExceptionFilter implements ExceptionFilter {
  catch(exception: HttpException, host: ArgumentsHost) {
    const ctx = host.switchToHttp();
    const response = ctx.getResponse<Response>();
    const request = ctx.getRequest<Request>();
    const status = exception.getStatus();

    response
      .status(status)
      .json({
        statusCode: status,
        timestamp: new Date().toISOString(),
        path: request.url,
      });
  }
}
```

## 绑定过滤器

将`HttpExceptionFileter` 绑定到 `CatsController` 的 `create()`方法上。

cats.controller.ts

~~~ ts
@Post()
@UseFilters(new HttpExceptionFilter())
async create(@Body() createCatDto:CreateCatDto) {
    throw new ForbiddenException();
}
~~~

>   `@UseFilters()`装饰器需要从‘@nestjs/common’包中导入

另一种方式时传递类(不是实例)，让框架承担实例化责任并启用依赖注入。

```typescript
@Post()
@UseFilters(HttpExceptionFilter)
async create(@Body() createCatDto: CreateCatDto) {
  throw new ForbiddenException();
}
```

>   尽可能使用类而不是使用实例。由于nest可以轻松地在整个模块中重复使用同一个实例，因此可以减少内存使用。

#### 控制器范围的过滤器

~~~ ts
@UseFilters(new HttpExceptionFilter())
export class CatsController {}
~~~

#### 全局范围过滤器

```typescript
async function bootstrap() {
  const app = await NestFactory.create(AppModule);
  app.useGlobalFilters(new HttpExceptionFilter());
  await app.listen(3000);
}
bootstrap();
```

>   `useGlobalFilters()`方法不会为网关和混合应用程序设置过滤器

使用这种方法，不能注入依赖，因为它不属于任何一个模块。为了解决这个问题，可以注册一个全局的过滤器直接为任何模块设置过滤器

app.module.ts

~~~ ts
import { Module } from '@nestjs/common';
import { APP_FILTER } from '@nestjs/core';

@Module({
  providers: [
    {
      provide: APP_FILTER,
      useClass: HttpExceptionFilter,
    },
  ],
})
export class AppModule {}
~~~

## Catch() 异常捕获

为了捕获每一个未处理的异常(不管异常类型如何)，将`@Catch()`装饰器的参数设为空。

```typescript
import {
  ExceptionFilter,
  Catch,
  ArgumentsHost,
  HttpException,
  HttpStatus,
} from '@nestjs/common';

@Catch()
export class AllExceptionsFilter implements ExceptionFilter {
  catch(exception: unknown, host: ArgumentsHost) {
    const ctx = host.switchToHttp();
    const response = ctx.getResponse();
    const request = ctx.getRequest();

    const status =
      exception instanceof HttpException
        ? exception.getStatus()
        : HttpStatus.INTERNAL_SERVER_ERROR;

    response.status(status).json({
      statusCode: status,
      timestamp: new Date().toISOString(),
      path: request.url,
    });
  }
}
```

这将捕获抛出的每一个异常，而不管其类型如何

## 继承

创建自定义的异常过滤器，如果希望重用已经实现的核心异常过滤器，并基于某些因素重写行为。

```typescript
import { Catch, ArgumentsHost } from '@nestjs/common';
import { BaseExceptionFilter } from '@nestjs/core';

@Catch()
export class AllExceptionsFilter extends BaseExceptionFilter {
  catch(exception: unknown, host: ArgumentsHost) {
    super.catch(exception, host);
  }
}
```

>   继承自基础类的过滤器必须由框架本身实例化（不要使用 `new` 关键字手动创建实例）